File: wrench-app/src/main/java/com/izettle/wrench/dialogs/scope/ScopeFragmentViewModel.java
package com.izettle.wrench.dialogs.scope;

import android.database.sqlite.SQLiteException;

import com.izettle.wrench.database.WrenchScope;
import com.izettle.wrench.database.WrenchScopeDao;

import java.util.Date;
import java.util.List;

import javax.inject.Inject;

import androidx.annotation.Nullable;
import androidx.annotation.WorkerThread;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.ViewModel;

public class ScopeFragmentViewModel extends ViewModel {
    private long applicationId;
    private LiveData<WrenchScope> selectedScopeLiveData;
    private WrenchScope selectedScope;
    private WrenchScopeDao scopeDao;

    @SuppressWarnings("WeakerAccess")
    @Inject
    public ScopeFragmentViewModel(WrenchScopeDao scopeDao) {
        this.scopeDao = scopeDao;
    }

    public void init(long applicationId) {
        this.applicationId = applicationId;
    }

    LiveData<List<WrenchScope>> getScopes() {
        return scopeDao.getScopes(applicationId);
    }

    void selectScope(WrenchScope wrenchScope) {
        wrenchScope.setTimeStamp(new Date());
        scopeDao.update(wrenchScope);
    }

    @WorkerThread
    WrenchScope createScope(String scopeName) throws SQLiteException {
        WrenchScope wrenchScope = new WrenchScope();
        wrenchScope.setName(scopeName);
        wrenchScope.setApplicationId(applicationId);
        wrenchScope.setId(scopeDao.insert(wrenchScope));

        return wrenchScope;
    }

    @WorkerThread
    void removeScope(WrenchScope scope) {
        scopeDao.delete(scope);
    }

    LiveData<WrenchScope> getSelectedScopeLiveData() {
        if (selectedScopeLiveData == null) {
            selectedScopeLiveData = scopeDao.getSelectedScopeLiveData(applicationId);
        }
        return selectedScopeLiveData;
    }

    @Nullable
    WrenchScope getSelectedScope() {
        return selectedScope;
    }

    void setSelectedScope(WrenchScope scope) {
        this.selectedScope = scope;
    }

}

File: wrench-app/src/main/java/com/izettle/wrench/dialogs/scope/ScopeFragmentViewModel.kt
package com.izettle.wrench.dialogs.scope

import android.database.sqlite.SQLiteException
import androidx.lifecycle.LiveData
import androidx.lifecycle.ViewModel
import com.izettle.wrench.database.WrenchScope
import com.izettle.wrench.database.WrenchScopeDao
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import java.util.*

class ScopeFragmentViewModel
constructor(private val scopeDao: WrenchScopeDao) : ViewModel() {
    private var applicationId: Long = 0

    internal val selectedScopeLiveData: LiveData<WrenchScope> by lazy {
        scopeDao.getSelectedScopeLiveData(applicationId)
    }

    internal var selectedScope: WrenchScope? = null

    internal val scopes: LiveData<List<WrenchScope>>
        get() = scopeDao.getScopes(applicationId)

    fun init(applicationId: Long) {
        this.applicationId = applicationId
    }

    internal fun selectScope(wrenchScope: WrenchScope) {
        GlobalScope.launch {
            wrenchScope.timeStamp = Date()
            scopeDao.update(wrenchScope)
        }
    }

    @Throws(SQLiteException::class)
    internal fun createScope(scopeName: String) {
        GlobalScope.launch {
            val wrenchScope = WrenchScope()
            wrenchScope.name = scopeName
            wrenchScope.applicationId = applicationId
            wrenchScope.id = scopeDao.insert(wrenchScope)
        }
    }

    internal fun removeScope(scope: WrenchScope) {
        GlobalScope.launch {
            scopeDao.delete(scope)
        }
    }
}
